flexmeasures/data/services/sensors.py

```diff
@@ -56,34 +57,53 @@ def get_sensors(
     return db.session.scalars(sensor_query).all()
 
 
-def _get_sensor_bdf(sensor: Sensor, staleness_search: dict) -> BeliefsDataFrame | None:
-    """Get bdf for a given sensor with given search parameters."""
+def _get_sensor_bdfs_by_source(
+    sensor: Sensor, staleness_search: dict
+) -> DataFrameGroupBy | None:
+    """Get bdf split by source for a given sensor with given search parameters."""
     bdf = TimedBelief.search(
         sensors=sensor,
         most_recent_events_only=True,
         **staleness_search,
     )
     if bdf.empty:
         return None
-    return bdf
+    return bdf.groupby(level=["source"], group_keys=False)
 
 
-def get_most_recent_knowledge_time(
-    sensor: Sensor, staleness_search: dict
-) -> datetime | None:
-    """Get the knowledge time of the sensor's most recent event.
+def get_staleness_start_times(
+    sensor: Sensor, staleness_search: dict, now: datetime
+) -> Dict[str, timedelta] | None:
+    """Get staleness start times for a given sensor by source.
+    For scheduler and forecaster sources staleness start is latest event start time.
 
+    For other sources staleness start is the knowledge time of the sensor's most recent event.
     This knowledge time represents when you could have known about the event
     (specifically, when you could have formed an ex-post belief about it).
     """
-    staleness_bdf = _get_sensor_bdf(sensor=sensor, staleness_search=staleness_search)
-    return None if staleness_bdf is None else staleness_bdf.knowledge_times[-1]
+    staleness_bdfs = _get_sensor_bdfs_by_source(
+        sensor=sensor, staleness_search=staleness_search
+    )
+    if staleness_bdfs is None:
+        return None
+
+    start_times = dict()
+    for source, bdf in staleness_bdfs:
+        time_column = "knowledge_times"
+        source = str(source)
+        if source in ("scheduler", "forecaster"):
```

Comment:
This should be `source.type`. The source name could be anything, but we have same specific types that we'd like to distinguish.